### DISCUSSION TITLE
fastapi dev runs stuck

### DISCUSSION THREAD

**Discussion by kingcwt** [fastapi/fastapi#14514] [Questions] [answered]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

The first time I ran fastapi dev, there was no problem, but I encountered a code error. Then after I modified the code and saved it, a WARNING: WatchFiles detected changes in 'tool/index.py'. Reloading... then got stuck and could not exit.

<img width="983" alt="image" src="https://github.com/tiangolo/fastapi/assets/29888469/f596a4fb-af29-4098-b489-3d9efee08f2c">

**Comment by ankit-brijwasi** [comment]
Any updates on this @kingcwt? facing the same issue!

**Comment by 3to1null** [comment]
Could be related to: https://github.com/encode/uvicorn/issues/2000

**Comment by Omachonu-prosper** [comment]
I currently have a slightly different issue from yours @kingcwt but it's still an issue with fastapi cli (dev and run)
The Code 
![image](https://github.com/tiangolo/fastapi/assets/69792620/c51f8e7a-79ab-437c-883b-8357ae074b35)


The Error
![image](https://github.com/tiangolo/fastapi/assets/69792620/82c3bbf7-903f-4c8c-b0f4-9b0caaba0c21)
![image](https://github.com/tiangolo/fastapi/assets/69792620/0ef1037d-7e28-4bd0-ade1-56b7307871c7)

**Comment by AIdjis** [comment]
do not use the fastapi dev to run your application just use uvicorn  instead, i do not face any issue with uvicorn

**Comment by ceb10n** [comment, marked as answer]
I tryied to reproduce it, but without success.

But sometimes I get the same problem running uvicorn with --reload in cmder/windows. I think it is related to uvicorn instead of fastapi

**Comment by RasmusN** [comment]
I'm having the same issue here on windows 10. 
I think this is related: https://www.reddit.com/r/FastAPI/comments/13oynf6/keeps_old_routes_after_changing_cache_problem/

It is like fastapi keeps an old version of the code cached somewhere. Even if I force kill fastpi and restart it, it loads an old version of my code.

It works (for a while) if I change port.

edit: this seems to be a uvicorn [open issue](https://github.com/encode/uvicorn/issues/1972) which in turn is caused by a bug in cpython ([open issue](https://github.com/python/cpython/issues/80116))

**Comment by Sarmad426** [comment]
I think uvicorn dev server is better as of now. As we are not getting any updates about the Fast API cli. Its important to note that this may be fixed in near future. It happens a lot with Fast API server. So you have to kill the terminal and open a new one to re-run the project. Its painful.

**Comment by arslanahmad768** [comment]
@Sarmad426 Yes, you are right i am also facing the same issue I tried with both the commands such as 
fast API dev filename 
or 
if __name__ == "__main__":
    uvicorn.run("main:app", host="127.0.0.1", port=8000, reload=True)
Sometimes it stuck you have to forcefully close the server but still, the port is running on the cache. So you have to run the server again on a different port.

**Comment by thecouchcoder** [comment]
Running into a similar issue.  I can I kill the the process with ctrl+c, but I cannot run on the same port.  Attempting to run on the same port prints out the message like everything is working fine, but trying to access /docs spins forever.  Switching to a new port works fine.

- Windows PowerShell (occurs both from Windows Terminal and PyCharm Terminal)
- Python 3.12
- Fast API 0.112.2

**Comment by WilliamStam** [comment]
in pycharm if you use the "run" method to run the project then it tends to hand after reloading. if you task manager kill python it reloads again. something definately fishy there. 

if you just do uvicorn --reload  in terminal it tends to work better but still a massive PITA on windows :( easier to just not run it with reload 

win11, py3.12, pycharm

**Comment by rutexd** [comment]
can confirm. win 11. 
issue present at least month. 

```WARNING:  WatchFiles detected changes in 'config.py'. Reloading...``` and its stuck. 
Its works fine until certain moment but hard to debug when exactly. after kill and start new it can stuck again.

No pycharm, vscode + .venv, server started by hand in terminal (❯ fastapi dev webserver.py --port 8080)

UPDATE:

> do not use the fastapi dev to run your application just use uvicorn instead, i do not face any issue with uvicorn

relating to the comments, this doesnt look to help aswell. 
![image](https://github.com/user-attachments/assets/da8d07e1-cb27-4e33-babc-3a99c1c237c6)

**Comment by mattmess1221** [comment]
I wonder if this would still happen with uvicorn's `--timeout-graceful-shutdown` option. Using `fastapi dev`, you have to set it using an environment variable `UVICORN_TIMEOUT_GRACEFUL_SHUTDOWN=5`.

**Reply by gaardhus** [reply]
For me it worked when doing

```
uv run uvicorn app.main:app --reload --timeout-graceful-shutdown 5
```
or

```
UVICORN_TIMEOUT_GRACEFUL_SHUTDOWN=5 uv run uvicorn app.main:app --reload
```
But not 

```
UVICORN_TIMEOUT_GRACEFUL_SHUTDOWN=5 uv run fastapi dev app/main.py
```

**Comment by atomicjarrod** [comment]
I have been getting this same problem whether `uvicorn main:app --reload` or `fastapi dev` method launching from powershell. (no pycharm here)

- Win11. 
- VSCode
- Python 3.12.6
- FastAPI 0.115.0
- Uvicorn 0.31.0

Resorting to Task Manager (Ctrl+Shift+Esc) and killing python tasks as Ctrl+C often won't quit server and closing powershell window and launching another doesn't seem to kill it either.

Installed fastapi using `"fastapi[standard]"` and also removed and tried just installing separately `fastapi` and `uvicorn`
Based on this thread [#2000](https://github.com/encode/uvicorn/issues/2000) just to test tried downgrading just uvicorn to 0.22 with no difference then to 0.21 (0.21.1) and sure enough it stops being stuck and reloads in a second?

```
WARNING:  StatReload detected changes in 'api.py'. Reloading...
INFO:     Started server process [9192]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```

**Comment by pmsmanishchauhan** [comment]
I am also  facing this issue recently with **FastAPI** reloading.
When we start the server, it works fine, and the APIs are functioning correctly. 
However, when we make changes to the code and save it, the FastAPI server does not respond when the API is hit again, and the FastAPI server page shows the reload icon for a long time.

**Comment by Jkrox** [comment]
Windows stuff, I see, using WSL2 never encountered this issue.

**Comment by qodzero** [comment]
For me the problem was avast scanning the app before allowing it to execute thus taking a long time. Disabled avast and suddenly it was working well

**Comment by pmsmanishchauhan** [comment]
![Screenshot 2024-10-09 115700](https://github.com/user-attachments/assets/29486dc4-1aa7-4fc4-bc69-beca095130d7)
I am facing a similar issue using Uvicorn. When I make changes to the code and save it, Uvicorn reloads, showing the message 'Application shutdown completed.' After that, the APIs stop responding, and the terminal becomes unresponsive. Even pressing Ctrl+C doesn't work

**Comment by AdmWinther** [comment]
I had the same problem with PyCharm FastAPI. When introducing a change, it used to get stuck on reload. (most likely couldn't kill the current process.) But I could resolve it by just using the terminal and running
```
uvicorn main:app --reload
```

It then worked like a charm.

I noticed that no matter what I do in Run/Debug Configuration of PyCharm, it wont be fixed there. I.E. if I choose Uvicorn in Run/Debug Config, I would still have the same problem. For me it works only when I lunch the server from terminal using Uvicorn.

**Comment by schrodienieur** [comment]
I had same problem, `fastapi dev` got stuck. Then after i killed the terminal and started again, the schemas in the docs got stucks, it won't be updated even i modify the model schemas, it will works again when you start dev server with different port.

**Comment by me-v2** [comment]
how fix it>

**Comment by me-v2** [comment]
backto uvicorn ==0.20.0,  this v can work.

**Comment by pmsmanishchauhan** [comment]
> backto uvicorn ==0.20.0, this v can work.

I have tried this method too, but the issue persists.

**Comment by me-v2** [comment]
> > backto uvicorn ==0.20.0, this v can work.
> 
> I have tried this method too, but the issue persists.

down 0.19, try again.

**Comment by ahturno** [comment]
> ### Privileged issue
> * [x]  I'm [@tiangolo](https://github.com/tiangolo) or he asked me directly to create an issue here.
> 
> ### Issue Content
> The first time I ran fastapi dev, there was no problem, but I encountered a code error. Then after I modified the code and saved it, a WARNING: WatchFiles detected changes in 'tool/index.py'. Reloading... then got stuck and could not exit.
> 
> <img alt="image" width="983" src="https://private-user-images.githubusercontent.com/29888469/329907854-f596a4fb-af29-4098-b489-3d9efee08f2c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkxMjA4NTEsIm5iZiI6MTczOTEyMDU1MSwicGF0aCI6Ii8yOTg4ODQ2OS8zMjk5MDc4NTQtZjU5NmE0ZmItYWYyOS00MDk4LWI0ODktM2Q5ZWZlZTA4ZjJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA5VDE3MDIzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjOWI4MWE5NTQyMGIxYjNiZTk2YzQ1YmYwY2E0MDZlZDEwMjQzMmRlZWIyNDczMmJiYzA2NTgwZjBjNTc4MTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.i1xt1sHsnxjD3QOrXM1jSSOSrdg-ggANnW4rR4m008o">


This issue appears to affect only the latest version of Uvicorn on Windows machines. It seems that the latest Uvicorn release may not be fully compatible with the Windows environment.  

A potential workaround is to downgrade to an earlier version of Uvicorn. In my case, version **0.20.0** runs smoothly on Windows 11 without any issues.  

To downgrade, follow these steps:  

1. Uninstall the current version of Uvicorn:  

   ```bash
   pip uninstall uvicorn

2. Install Uvicorn version 0.20.0: 

   ```bash
   pip install uvicorn==0.20.0
   ```

This solution worked for me, and it may help others experiencing the same issue. Let me know if this resolves the problem!

**Comment by Kludex** [comment]
Don't install a version that is more than 2 years old, please.

**Comment by WilliamStam** [comment]
> Don't install a version that is more than 2 years old, please.

we had to do the same. in pycharm its nearly impossible to dev on anything after that. strange that its happening in vscode. 

https://youtrack.jetbrains.com/issue/PY-60962/os.killpid-signal.CRTLCEVENT-ignored-when-running-python-program-from-Pycharm

**Comment by Pa1ochka** [comment]
If someone has found a solution, please let me know.

**Comment by zhlee1997** [comment]
Is there any solution for this issue?

**Comment by Pa1ochka** [comment]
I made a .bat file that terminates all python processes.It's not very convenient, but it works.I have to restart it every time.
Something like that:

taskkill /f /im python.exe
taskkill /f /im pythonw.exe

**Comment by chmielcode** [comment]
> I wonder if this would still happen with uvicorn's `--timeout-graceful-shutdown` option. Using `fastapi dev`, you have to set it using an environment variable `UVICORN_TIMEOUT_GRACEFUL_SHUTDOWN=5`.

I 'm on uvicorn@0.34.0 and it worked like a charm :) Tested multiple times by adding and removing this to command line: uvicorn [other args] _--timeout-graceful-shutdown 1_
Thank you!

**Comment by leavor** [comment]
pip install uvicorn==0.21.1

**Comment by rxy1212** [comment]
Is the "fastapi run" command production-ready?

**Comment by WilliamStam** [comment]
just an interesting observation. i suspect this has more to do with pycharm "run" terminal not behaving correctly. 

built a test loop thing with a try catch on keyboardinterupt and if run through "run" it never ends the loop. but if run in cmd prompt (or even pycharms normal terminal) it works fine. (while True sleep(1) type thing)

its been many years... there are a ton of jetbrains tickets for this i think. 

we moved on to using the newer versions of uvicorn now again but when dev'ing we physically stop the fast api and restart it to test stuff (so not really using --reload any more). - 

click run. click stop twice. click run. 

<img width="152" height="51" alt="Image" src="https://github.com/user-attachments/assets/7024b125-d2e3-455b-90c7-1d7a5b462d27" />

**Comment by emrecpp** [comment]
I can’t believe this issue still exists. I also can’t believe that installing such an old version, uvicorn==0.20.0, fixes it.

**Comment by tiangolo** [comment]
It seems this was a question, it should have been started in GitHub Discussions, so I'll move it.

It seems this is not related to FastAPI or FastAPI CLI but Uvicorn, and maybe not even that but PyCharm, it should probably be asked there.

**Comment by WilliamStam** [comment]
this seems fixed now. for the sake of the discussion. we started updating our uvicorns to the latest now (recent within the month something in pycharm changed). can anyone else confirm that it works? it still freaks all of us in the co its that new....
